Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -95,68 +103,170 @@ internal SFSession GetSession(string connStr, SecureString password)
         {
             s_logger.Debug("SessionPool::GetSession");
             if (!_pooling)
-                return NewSession(connStr, password);
-            SFSession session = GetIdleSession(connStr);
-            return session ?? NewSession(connStr, password);
+                return NewSession(connStr, password, _noPoolingSessionCreationTokenCounter.NewToken());
+            var sessionOrCreateToken = GetIdleSession(connStr);
+            if (sessionOrCreateToken.Session != null)
+            {
+                _sessionPoolEventHandler.OnSessionProvided(this);
+            }
+            return sessionOrCreateToken.Session ?? NewSession(connStr, password, sessionOrCreateToken.SessionCreationToken);
         }
         
-        internal Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
+        internal async Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
         {
             s_logger.Debug("SessionPool::GetSessionAsync");
             if (!_pooling)
-                return NewSessionAsync(connStr, password, cancellationToken);
-            SFSession session = GetIdleSession(connStr);
-            return session != null ? Task.FromResult(session) : NewSessionAsync(connStr, password, cancellationToken);
+                return await NewSessionAsync(connStr, password, _noPoolingSessionCreationTokenCounter.NewToken(), cancellationToken).ConfigureAwait(false);
+            var sessionOrCreateToken = GetIdleSession(connStr);
+            if (sessionOrCreateToken.Session != null)
+            {
+                _sessionPoolEventHandler.OnSessionProvided(this);
+            }
+            return sessionOrCreateToken.Session ?? await NewSessionAsync(connStr, password, sessionOrCreateToken.SessionCreationToken, cancellationToken).ConfigureAwait(false);
         }
 
         internal SFSession GetSession() => GetSession(ConnectionString, Password);
 
         internal Task<SFSession> GetSessionAsync(CancellationToken cancellationToken) =>
             GetSessionAsync(ConnectionString, Password, cancellationToken);
 
-        private SFSession GetIdleSession(string connStr)
+        internal void SetSessionPoolEventHandler(ISessionPoolEventHandler sessionPoolEventHandler)
+        {
+            _sessionPoolEventHandler = sessionPoolEventHandler;
+        }
+        
+        private SessionOrCreationToken GetIdleSession(string connStr)
         {
             s_logger.Debug("SessionPool::GetIdleSession");
             lock (_sessionPoolLock)
             {
-                for (int i = 0; i < _idleSessions.Count; i++)
+                if (_waitingForSessionToReuseQueue.IsAnyoneWaiting())
+                {
+                    s_logger.Debug("SessionPool::GetIdleSession - someone is already waiting for a session, request is going to be queued");
```

Comment:
Shouldn't we first check what happens in ln 156 and create session if possible rather get to the waiting queue. It seems to be quicker way to get a new session.